src/v/iceberg/conversion/values_protobuf.cc

```diff
@@ -232,6 +235,164 @@ convert_timestamp(std::unique_ptr<parsed::message> message) {
     co_return value_outcome{iceberg::timestamp_value{absl::ToUnixMicros(ts)}};
 }
 
+// Forward declaration for recursive calls
+ss::future<result<std::monostate, value_conversion_exception>>
+serialize_protobuf_value_to_json(
+  serde::json::writer& writer, const parsed::message& value_msg, int depth);
+
+ss::future<result<std::monostate, value_conversion_exception>>
+serialize_protobuf_list_to_json(
+  serde::json::writer& writer, const parsed::message& list, int depth) {
+    if (depth >= max_recursion_depth) {
+        co_return value_conversion_exception(
+          fmt::format(
+            "Maximum recursion depth {} exceeded", max_recursion_depth));
+    }
+
+    writer.begin_array();
+    if (auto it = list.fields.find(1); it != list.fields.end()) {
+        const auto& repeated_data = std::get<parsed::repeated>(it->second);
+        const auto& messages
+          = std::get<chunked_vector<std::unique_ptr<parsed::message>>>(
+            repeated_data.elements);
+
+        for (const auto& msg_ptr : messages) {
+            auto result = co_await serialize_protobuf_value_to_json(
+              writer, *msg_ptr, depth + 1);
+            if (result.has_error()) {
+                co_return result.error();
+            }
+        }
+    }
+    writer.end_array();
+    co_return std::monostate{};
+}
+
+ss::future<result<std::monostate, value_conversion_exception>>
+serialize_protobuf_map_to_json(
+  serde::json::writer& writer, const parsed::message& map, int depth) {
+    if (depth >= max_recursion_depth) {
+        co_return value_conversion_exception(
+          fmt::format(
+            "Maximum recursion depth {} exceeded", max_recursion_depth));
+    }
+
+    writer.begin_object();
+    if (auto it = map.fields.find(1); it != map.fields.end()) {
+        const auto& map_data = std::get<parsed::map>(it->second);
+        for (const auto& [key, value] : map_data.entries) {
+            // google.protobuf.Struct maps always have string keys
+            writer.key(std::get<iobuf>(key));
+            // Convert value
+            const auto& value_msg_ptr
+              = std::get<std::unique_ptr<parsed::message>>(value);
+            auto result = co_await serialize_protobuf_value_to_json(
+              writer, *value_msg_ptr, depth + 1);
+            if (result.has_error()) {
+                co_return result.error();
+            }
+        }
+    }
+    writer.end_object();
+    co_return std::monostate{};
+}
+
+ss::future<result<std::monostate, value_conversion_exception>>
+serialize_protobuf_value_to_json(
+  serde::json::writer& writer, const parsed::message& value_msg, int depth) {
+    if (depth >= max_recursion_depth) {
+        co_return value_conversion_exception(
+          fmt::format(
+            "Maximum recursion depth {} exceeded", max_recursion_depth));
+    }
+
+    for (const auto& [field_num, field_value] : value_msg.fields) {
```

Comment:
Where are these magic numbers defined? Would be great if we could give them names somewhere

I'm kind of surprised this is a loop over fields, where the field num is a type, rather than some part of the descriptor. For my own understanding, what does it mean for value_msg.fields to have more than one field? Should that be considered bad data, considering we don't do anything with the other fields?

Reply:
See https://buf.build/protocolbuffers/wellknowntypes/file/main:google/protobuf/struct.proto#L63

That is the proto here, it's a oneof, so the proto parser ensures at most one value is set. We don't loop over a descriptor because it's hard coded the descriptor (being well known) and the point is it's json and dynamically typed